### ISSUE TITLE
Chapter 4. Classes and Interfaces

### ISSUE THREAD

**Issue by eziceice** [eziceice/EffectiveJava#3]

**Comment by eziceice** [comment]
## In public classes, use accessor methods, not public fields

 - If a class is accessible outside its package, provide accessor methods to preserve the flexibility to change the class's internal representation. Even if a class is package-private or is a private nested class, there is nothing inherently wrong with exposing its data fields.

**Comment by eziceice** [comment]
## Minimize mutability

 - An immutable class is simply a class whose instances cannot be modified. Example in Java: *String, BigInteger and BigDecimal*. 
 - There are five rules to make a class immutable
	 - Don't provide methods that modify the object's state.
	 - Ensure that the class can't be extended. (Generally making the class final, **or make constructor private and provide static factory method - Best alternative**.)
	 - Make all fields final. 
	 - Make all fields private. 
	 - Ensure exclusive access to any mutable components. (If your class has any fields that refer to mutable objects, ensure that clients of the class cannot obtain references to these objects.)
 - Immutable objects are simple, an immutable object can be in exactly one state, the state in which it was created.  Meanwhile, immutable objects are inherently thread-safe, they require no synchronization, and it can be shared freely. 
 - Not only can you share immutable objects, but they can share their internals.
 - Immutable objects make great building blocks for other objects. For example, by making it as a key in map or set. 
 - Immutable objects provide failure atomicity for free. 
 - The **major disadvantage** of immutable classes is that they require a separate object for each distinct value. The package-private mutable companion class approach works fine if you accurately predict which complex operations clients will want to perform on your immutable class. If not, then your best bet is to provide a public mutable companion class. For example, *String and StringBuilder*. 
 - In truth, no method may produce an externally visible change in the object's state. However, some immutable classes have one or more nonfinal fields in which they cache the results of expensive computations the first time they are needed. If the same value is requested again, the cached value is returned, saving the cost of recalculation.  

**In conclusion, classes should be immutable unless there's a very good reason to make them mutable. If a class cannot be made immutable, limit its mutability as much as possible and declare every field private final unless there's a good reason to do otherwise. Constructors should create fully initialized objects with all of their invariants established.**

**Comment by eziceice** [comment]
## Favor composition over inheritance

 - Unlike method invocation, inheritance violates encapsulation. 
 - **当子类重写了父类的函数，那么子类的对象如果调用该函数，一定调用的是重写过后的函数**

**Comment by eziceice** [comment]
## Design and document for inheritance or else prohibit it 

 - The class must document precisely the effects of overriding any method. In other words, the class must document its self-use of overridable methods. For each public or protected method, the documentation must indicate which overridable methods the method invokes, in what sequence and how the results of each invocation affect subsequent processing. 
 - More generally, a class must document any circumstances under which it might invoke an overridable method. @implSpec - Implementation Requirements can be used to document the inner workings of the method. 
 - A class may have to provide hooks into its internal workings in the form of judiciously chosen protected methods or, in rare instances, protected fields. 
 - The only way to test a class designed for inheritance is to write subclasses. 
 - Constructors must not invoke overridable methods, because the superclass constructor runs before the subclass constructor, so the overriding method in the subclass will get invoked before the subclass constructor has run. Same thing need to be considered when implementing Cloneable or Seriablizable, becasue clone or readObject method do the same thing as constructor. **Neither clone nor readObject may invoke an overrdable method, directly or indirectly.** 
```
public class Super {  
// Broken - constructor invokes an overridable method  
  public Super() {  
        overrideMe();  
  }  
    public void overrideMe() { 
  }  
}  
// Here’s a subclass that overrides the overrideMe method, which is erroneously invoked by Super’s sole constructor:  
public final class Sub extends Super {   
  private final Instant instant; // Blank final, set by constructor 
  Sub() {  
        instant = Instant.now();  
  }  
// Overriding method invoked by superclass constructor  
  @Override 
  public void overrideMe() {  
        System.out.println(instant);  
  }  
  public static void main(String[] args) {  
        Sub sub = new Sub();  
	sub.overrideMe(); 
	// one null - cuz the subclass has not been initialized 
	// one currenty time
  }  
}
```
 - In summary, designing a class for inheritance is hard to work. All of its self-use patterns must be documented, and once you've documented them, you must commit to them for the life of the class.

**Comment by eziceice** [comment]
## Prefer interfaces to abstract classes 

 - A skeletal implementation classes are called AbstractInterface, which is an abstract class implements some interface. (Example in Java - set, abstractSet, map and abstractMap.)
```
public interface Foo {
    void bar();
    void baz();
}

// Abstract Interface
private abstract class AbstractFoo implements Foo {
    ...
}
```

 - For most implementors of an interface with a skeletal implementation class, extending this class is the obvious choice, but it is strictly optional. If a class cannot be made to extend the skeletal implementation, the class can always implement the interface directly.

**Comment by eziceice** [comment]
## Design interfaces for posterity 

 - The declaration for a default method includes a default implementation that is used by all classes that implement the interface but do not implement the default method. While the addition of default methods to Java makes it possible to add methods to an existing interface, there is no guarantee that these methods will work in all preexisting implementations. Defaults methods are "injected" into existing implementations without the knowledge or consent of their implementors.

**Comment by eziceice** [comment]
## Favor static member classes over nonstatic 

 - A nested class should exist only to serve its enclosing class. If a nested class would be useful in some other context, then it should be a top-level class. There are four kinds of nested classes: static member classes, nonstatic member classes, anonymous classes, and local classes. **All but the first kind are known as inner classes**. 
 - A common use of a static member class is as a public helper class, useful only in conjunction with its outer class. 
 - One common use of a nonstatic member class is to define an Adapter that allows an instance of the outer class to be viewed as an instance of some unrelated class. 
 - If you declare a member class that does not require access to an enclosing instance, always put the static modifier in its declaration. If you omit this modifier, each instance will have a hidden extraneous reference to its enclosing instance, storing this reference takes time and space. More seriously, it can result in the enclosing instance being retained when it would otherwise be eligible for garbage collection.
